File: takengine/mapengine/sdk/src/feature/AbstractFeatureDataStore2.cpp
#include "feature/AbstractFeatureDataStore2.h"

#include <regex>

#include "feature/FeatureCursor2.h"
#include "thread/Lock.h"
#include "util/Logging.h"

using namespace TAK::Engine::Feature;

using namespace TAK::Engine::Thread;
using namespace TAK::Engine::Util;

using namespace atakmap::util;

namespace
{
    std::string replace(const std::string &src, const char old, const char *cnew)
    {
        std::string dst = src;
        std::size_t pos = 0u;
        std::size_t repLen = strlen(cnew);
        do 
        {
            pos = dst.find(old, pos);
            if (pos == std::string::npos)
                break;
            dst = dst.replace(pos, 1, cnew);
            pos += repLen;
        } while (true);
        return dst;
    }

    std::string replace(const std::string &src, const char *old, const char *cnew)
    {
        const std::size_t oldLen = strlen(old);
        std::string dst = src;
        std::size_t pos = 0u;
        std::size_t repLen = strlen(cnew);
        do
        {
            pos = dst.find(old, pos);
            if (pos == std::string::npos)
                break;
            dst = dst.replace(pos, oldLen, cnew);
            pos += repLen;
        } while (true);
        return dst;
    }

    std::string regexEscapeChar(const char *s)
    {
        // note backslash is FIRST character
        static char regexChars[16] = { '\\', '[', ']', '^', '-',
                                       '&', '$', '?', '*', '+',
                                       '{', '}', ',', '(', ')',
                                       '|', };

        std::string retval(s);
        char escaped[3] {'\\', '\0', '\0'};
        for (std::size_t i = 0; i < 16; i++) {
            escaped[1] = regexChars[i];
            retval = replace(retval, regexChars[i], escaped);
        }

        return retval;
    }
        }

AbstractFeatureDataStore2::AbstractFeatureDataStore2(int modificationFlags_, int visibilityFlags_) :
    modificationFlags(modificationFlags_),
    visibilityFlags(visibilityFlags_),
    inBulkModification(0),
    contentChanged(false),
    mutex(TEMT_Recursive)
{}
    
TAKErr AbstractFeatureDataStore2::addOnDataStoreContentChangedListener(OnDataStoreContentChangedListener *l) NOTHROWS
{
    TAKErr code(TE_Ok);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    // XXX - check for throws
    this->contentChangedListeners.insert(l);
    return TE_Ok;
}

    
TAKErr AbstractFeatureDataStore2::removeOnDataStoreContentChangedListener(OnDataStoreContentChangedListener *l) NOTHROWS
{
    TAKErr code(TE_Ok);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);
    this->contentChangedListeners.erase(l);
    return TE_Ok;
}

void AbstractFeatureDataStore2::setContentChanged() NOTHROWS
{
    this->contentChanged = true;
}

void AbstractFeatureDataStore2::dispatchDataStoreContentChangedNoSync(bool force) NOTHROWS
{
    if (this->inBulkModification > 0)
        return;
    if (this->contentChanged || force) {
        this->contentChanged = false;

        std::set<OnDataStoreContentChangedListener *>::iterator it;
        for (it = this->contentChangedListeners.begin(); it != this->contentChangedListeners.end(); it++)
            (*it)->onDataStoreContentChanged(*this);
    }
}

    
TAKErr AbstractFeatureDataStore2::getVisibilitySettingsFlags(int *value) NOTHROWS
{
    *value = this->visibilityFlags;
    return TE_Ok;
}

    
TAKErr AbstractFeatureDataStore2::setFeatureVisible(const int64_t fid, const bool visible) NOTHROWS
{
    if ((this->visibilityFlags&VISIBILITY_SETTINGS_FEATURE) != VISIBILITY_SETTINGS_FEATURE)
        return TE_Unsupported;
    TAKErr code(TE_Ok);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);
    code = this->setFeatureVisibleImpl(fid, visible);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}
    
TAKErr AbstractFeatureDataStore2::setFeaturesVisible(const FeatureQueryParameters &params, const bool visible) NOTHROWS
{
    if ((this->visibilityFlags&VISIBILITY_SETTINGS_FEATURE) != VISIBILITY_SETTINGS_FEATURE)
        return TE_Unsupported;
    TAKErr code(TE_Ok);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);
    code = this->setFeaturesVisibleImpl(params, visible);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

TAKErr AbstractFeatureDataStore2::setFeatureSetVisible(const int64_t setId, const bool visible) NOTHROWS
{
    if ((this->visibilityFlags&VISIBILITY_SETTINGS_FEATURESET) != VISIBILITY_SETTINGS_FEATURESET)
        return TE_Unsupported;
    TAKErr code(TE_Ok);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);
    code = this->setFeatureSetVisibleImpl(setId, visible);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

TAKErr AbstractFeatureDataStore2::setFeatureSetsVisible(const FeatureSetQueryParameters &params, const bool visible) NOTHROWS
{
    if ((this->visibilityFlags&VISIBILITY_SETTINGS_FEATURESET) != VISIBILITY_SETTINGS_FEATURESET)
        return TE_Unsupported;
    TAKErr code(TE_Ok);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);
    code = this->setFeatureSetsVisibleImpl(params, visible);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

    
TAKErr AbstractFeatureDataStore2::getModificationFlags(int *value) NOTHROWS {
    *value = this->modificationFlags;
    return TE_Ok;
}

    
TAKErr AbstractFeatureDataStore2::beginBulkModification() NOTHROWS
{
    TAKErr code;
    code = this->checkModificationFlags(MODIFY_BULK_MODIFICATIONS);
    TE_CHECKRETURN_CODE(code);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);
    code = this->beginBulkModificationImpl();
    if (code == TE_Ok)
        this->inBulkModification++;
    return code;
}

    
TAKErr AbstractFeatureDataStore2::endBulkModification(const bool successful) NOTHROWS
{
    TAKErr code;
    code = this->checkModificationFlags(MODIFY_BULK_MODIFICATIONS);
    TE_CHECKRETURN_CODE(code);

    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    if (this->inBulkModification == 0)
        return TE_IllegalState;

    code = this->endBulkModificationImpl(successful);
    if (code == TE_Ok) {
        this->inBulkModification--;
        if (this->inBulkModification == 0)
            this->dispatchDataStoreContentChangedNoSync(false);
    }
    return code;
}

TAKErr AbstractFeatureDataStore2::isInBulkModification(bool *value) NOTHROWS
{
    TAKErr code(TE_Ok);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    *value = (this->inBulkModification > 0);
    return TE_Ok;
}

TAKErr AbstractFeatureDataStore2::insertFeatureSet(FeatureSetPtr_const *featureSet, const char *provider, const char *type, const char *name, const double minResolution, const double maxResolution) NOTHROWS
{
    TAKErr code;
    code = this->checkModificationFlags(MODIFY_FEATURESET_INSERT);
    TE_CHECKRETURN_CODE(code);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    code = this->insertFeatureSetImpl(featureSet, provider, type, name, minResolution, maxResolution);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

    
TAKErr AbstractFeatureDataStore2::updateFeatureSet(const int64_t fsid, const char *name) NOTHROWS
{
    TAKErr code;
    code = this->checkModificationFlags(MODIFY_FEATURESET_UPDATE | MODIFY_FEATURESET_NAME);
    TE_CHECKRETURN_CODE(code);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    code = this->updateFeatureSetImpl(fsid, name);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

    
TAKErr AbstractFeatureDataStore2::updateFeatureSet(const int64_t fsid, const double minResolution, const double maxResolution) NOTHROWS
{
    TAKErr code;
    code = this->checkModificationFlags(MODIFY_FEATURESET_UPDATE | MODIFY_FEATURESET_DISPLAY_THRESHOLDS);
    TE_CHECKRETURN_CODE(code);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    code = this->updateFeatureSetImpl(fsid, minResolution, maxResolution);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

    
TAKErr AbstractFeatureDataStore2::updateFeatureSet(const int64_t fsid, const char *name, const double minResolution, const double maxResolution) NOTHROWS
{
    TAKErr code;
    code = this->checkModificationFlags(MODIFY_FEATURESET_UPDATE | MODIFY_FEATURESET_NAME | MODIFY_FEATURESET_DISPLAY_THRESHOLDS);
    TE_CHECKRETURN_CODE(code);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    code = this->updateFeatureSetImpl(fsid, name, minResolution, maxResolution);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

    
TAKErr AbstractFeatureDataStore2::deleteFeatureSet(const int64_t fsid) NOTHROWS
{
    TAKErr code;
    code = this->checkModificationFlags(MODIFY_FEATURESET_DELETE);
    TE_CHECKRETURN_CODE(code);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    code = this->deleteFeatureSetImpl(fsid);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

    
TAKErr AbstractFeatureDataStore2::deleteAllFeatureSets() NOTHROWS
{
    TAKErr code;
    code = this->checkModificationFlags(MODIFY_FEATURESET_DELETE);
    TE_CHECKRETURN_CODE(code);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    code = this->deleteAllFeatureSetsImpl();
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

    
TAKErr AbstractFeatureDataStore2::insertFeature(FeaturePtr_const *feature, const int64_t fsid, const char *name, const atakmap::feature::Geometry &geom, const atakmap::feature::Style *style, const atakmap::util::AttributeSet &attributes) NOTHROWS
{
    TAKErr code;
    code = this->checkModificationFlags(MODIFY_FEATURESET_FEATURE_INSERT);
    TE_CHECKRETURN_CODE(code);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    code = this->insertFeatureImpl(feature, fsid, name, geom, style, attributes);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

    
TAKErr AbstractFeatureDataStore2::updateFeature(const int64_t fid, const char *name) NOTHROWS
{
    TAKErr code;
    code = this->checkModificationFlags(MODIFY_FEATURESET_FEATURE_UPDATE | MODIFY_FEATURE_NAME);
    TE_CHECKRETURN_CODE(code);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    code = this->updateFeatureImpl(fid, name);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

    
TAKErr AbstractFeatureDataStore2::updateFeature(const int64_t fid, const atakmap::feature::Geometry &geom) NOTHROWS
{
    TAKErr code;
    code = this->checkModificationFlags(MODIFY_FEATURESET_FEATURE_UPDATE | MODIFY_FEATURE_GEOMETRY);
    TE_CHECKRETURN_CODE(code);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    code = this->updateFeatureImpl(fid, geom);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

    
TAKErr AbstractFeatureDataStore2::updateFeature(const int64_t fid, const atakmap::feature::Style *style) NOTHROWS
{
    TAKErr code;
    code = this->checkModificationFlags(MODIFY_FEATURESET_FEATURE_UPDATE | MODIFY_FEATURE_STYLE);
    TE_CHECKRETURN_CODE(code);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    code = this->updateFeatureImpl(fid, style);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

    
TAKErr AbstractFeatureDataStore2::updateFeature(const int64_t fid, const atakmap::util::AttributeSet &attributes) NOTHROWS
{
    TAKErr code;
    code = this->checkModificationFlags(MODIFY_FEATURESET_FEATURE_UPDATE | MODIFY_FEATURE_ATTRIBUTES);
    TE_CHECKRETURN_CODE(code);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    code = this->updateFeatureImpl(fid, attributes);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

    
TAKErr AbstractFeatureDataStore2::updateFeature(const int64_t fid, const char *name, const atakmap::feature::Geometry &geom, const atakmap::feature::Style *style, const atakmap::util::AttributeSet &attributes) NOTHROWS
{
    TAKErr code;
    code = this->checkModificationFlags(MODIFY_FEATURESET_FEATURE_UPDATE | MODIFY_FEATURE_NAME | MODIFY_FEATURE_GEOMETRY | MODIFY_FEATURE_STYLE | MODIFY_FEATURE_ATTRIBUTES);
    TE_CHECKRETURN_CODE(code);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    code = this->updateFeatureImpl(fid, name, geom, style, attributes);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

    
TAKErr AbstractFeatureDataStore2::deleteFeature(const int64_t fid) NOTHROWS
{
    TAKErr code;
    code = this->checkModificationFlags(MODIFY_FEATURESET_FEATURE_DELETE);
    TE_CHECKRETURN_CODE(code);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    code = this->deleteFeatureImpl(fid);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

    
TAKErr AbstractFeatureDataStore2::deleteAllFeatures(const int64_t fsid) NOTHROWS
{
    TAKErr code;
    code = this->checkModificationFlags(MODIFY_FEATURESET_FEATURE_DELETE);
    TE_CHECKRETURN_CODE(code);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, mutex);
    TE_CHECKRETURN_CODE(code);

    code = this->deleteAllFeaturesImpl(fsid);
    if (code == TE_Ok)
        this->dispatchDataStoreContentChangedNoSync(false);
    return code;
}

TAKErr AbstractFeatureDataStore2::checkModificationFlags(const int capability) NOTHROWS {
    const bool retval = ((this->modificationFlags&capability) == capability);
    if (!retval)
        return TE_Unsupported;
    return TE_Ok;
}

TAKErr AbstractFeatureDataStore2::matches(bool *matched, const char *ctest, const char *value, const char wildcard) NOTHROWS
{
    // NULL cases
    if (!ctest && !value) {
        *matched = true;
        return TE_Ok;
    } else if (!ctest || !value) {
        *matched = false;
        return TE_Ok;
    }

    std::string test = ctest;
    if (test.find(wildcard) == std::string::npos) {
        *matched = (strcmp(value, ctest) == 0);
        return TE_Ok;
    }

    const char scratch[3] = { wildcard, wildcard, '\0' };

    while (test.find(scratch) != std::string::npos)
        test = replace(test, scratch, scratch+1);
    test = regexEscapeChar(test.c_str());
    test = replace(test, wildcard, ".*");
    try {
        *matched = std::regex_match(value, std::regex(test));
        return TE_Ok;
    } catch (std::regex_error &e) {
        Logger::log(Logger::Error, "AbstractFeatureDataStore2::matches: failed to construct regular expression from %s", test.c_str());
        return TE_Err;
    }
}

TAKErr AbstractFeatureDataStore2::matches(bool *matched, TAK::Engine::Port::Collection<TAK::Engine::Port::String> &test, const char *value, const char wildcard) NOTHROWS
{
    *matched = false;
    if (test.empty())
        return TE_Ok;

    TAKErr code;
    TAK::Engine::Port::Collection<TAK::Engine::Port::String>::IteratorPtr iter(NULL, NULL);
    code = test.iterator(iter);
    TE_CHECKRETURN_CODE(code);

    TAK::Engine::Port::String arg;
    do {
        code = iter->get(arg);
        if (code != TE_Ok)
            break;
        code = matches(matched, arg, value, wildcard);
        TE_CHECKBREAK_CODE(code);
        if (*matched)
            break;

        code = iter->next();
        if (code != TE_Ok)
            break;
    } while (true);

    return (code == TE_Done) ? TE_Ok : code;
}

TAKErr AbstractFeatureDataStore2::queryFeaturesCount(int *value, FeatureDataStore2 &dataStore, const FeatureQueryParameters &params) NOTHROWS
{
    TAKErr code;
    FeatureCursorPtr result(NULL, NULL);
    code = dataStore.queryFeatures(result, params);
    TE_CHECKRETURN_CODE(code);
    int v = 0;
    
    do {
        code = result->moveToNext();
        TE_CHECKBREAK_CODE(code);
        v++;
    } while (true);
    if (code == TE_Done)
        code = TE_Ok;
    TE_CHECKRETURN_CODE(code);
    if (value)
        *value = v;
    
    return code;
}

File: takengine/takengine/src/main/java/com/atakmap/map/layer/feature/AbstractFeatureDataStore2.java
package com.atakmap.map.layer.feature;

import java.util.Collection;
import java.util.Collections;
import java.util.IdentityHashMap;
import java.util.LinkedList;
import java.util.Set;

import com.atakmap.map.layer.feature.geometry.Geometry;
import com.atakmap.map.layer.feature.style.Style;
import com.atakmap.math.MathUtils;
import com.atakmap.util.WildCard;

public abstract class AbstractFeatureDataStore2 implements FeatureDataStore {

    protected Set<OnDataStoreContentChangedListener> contentChangedListeners;
    protected final int visibilityFlags;
    protected final int modificationFlags;

    private int inBulkModification;

    protected AbstractFeatureDataStore2(int modificationFlags, int visibilityFlags) {
        this.modificationFlags = modificationFlags;
        this.visibilityFlags = visibilityFlags;
        
        this.contentChangedListeners = Collections.newSetFromMap(new IdentityHashMap<OnDataStoreContentChangedListener, Boolean>()); 
    }

    protected void adopt(FeatureSet featureSet, long fsid, long version) {
        featureSet.id = fsid;
        featureSet.owner = this;
        featureSet.version = version;
    }

    @Override
    public synchronized void addOnDataStoreContentChangedListener(OnDataStoreContentChangedListener l) {
        this.contentChangedListeners.add(l);
    }

    @Override
    public synchronized void removeOnDataStoreContentChangedListener(OnDataStoreContentChangedListener l) {
        this.contentChangedListeners.remove(l);
    }

    protected void dispatchDataStoreContentChangedNoSync() {
        if(this.inBulkModification > 0)
            return;

        for(OnDataStoreContentChangedListener l : this.contentChangedListeners)
            l.onDataStoreContentChanged(this);
    }

    @Override
    public final int getVisibilitySettingsFlags() {
        return this.visibilityFlags;
    }

    @Override
    public final void setFeatureVisible(long fid, boolean visible) {
        if(!MathUtils.hasBits(this.visibilityFlags,VISIBILITY_SETTINGS_FEATURE))
            throw new UnsupportedOperationException();

        synchronized(this) {
            if(this.setFeatureVisibleImpl(fid, visible))
                this.dispatchDataStoreContentChangedNoSync();
        }
    }
    
    protected abstract boolean setFeatureVisibleImpl(long fid, boolean visible);

    @Override
    public final void setFeaturesVisible(FeatureQueryParameters params, boolean visible) {
        if(!MathUtils.hasBits(this.visibilityFlags,VISIBILITY_SETTINGS_FEATURE))
            throw new UnsupportedOperationException();
        synchronized(this) {
            if(this.setFeaturesVisibleImpl(params, visible))
                this.dispatchDataStoreContentChangedNoSync();
        }
    }
    
    protected abstract boolean setFeaturesVisibleImpl(FeatureQueryParameters params, boolean visible);
    
//    @Override
//    public abstract boolean isFeatureVisible(long fid);

    @Override
    public void setFeatureSetVisible(long setId, boolean visible) {
        if(!MathUtils.hasBits(this.visibilityFlags,VISIBILITY_SETTINGS_FEATURESET))
            throw new UnsupportedOperationException();
        synchronized(this) {
            if(this.setFeatureSetVisibleImpl(setId, visible))
                this.dispatchDataStoreContentChangedNoSync();
        }
    }
    
    protected abstract boolean setFeatureSetVisibleImpl(long setId, boolean visible);
    
    @Override
    public void setFeatureSetsVisible(FeatureSetQueryParameters params, boolean visible) {
        if(!MathUtils.hasBits(this.visibilityFlags,VISIBILITY_SETTINGS_FEATURESET))
            throw new UnsupportedOperationException();
        synchronized(this) {
            if(this.setFeatureSetsVisibleImpl(params, visible))
                this.dispatchDataStoreContentChangedNoSync();
        }
    }
    
    protected abstract boolean setFeatureSetsVisibleImpl(FeatureSetQueryParameters params, boolean visible);
    
    @Override
    public final int getModificationFlags() {
        return this.modificationFlags;
    }
    
    @Override
    public synchronized final boolean isInBulkModification() {
        return (this.inBulkModification > 0);
    }

    @Override
    public synchronized final void beginBulkModification() {
        this.checkModificationFlags(MODIFY_BULK_MODIFICATIONS, true);
        
        this.beginBulkModificationImpl();
        this.inBulkModification++;
    }
    
    protected abstract void beginBulkModificationImpl(); 
    
    @Override
    public synchronized final void endBulkModification(boolean successful) {
        this.checkModificationFlags(MODIFY_BULK_MODIFICATIONS, true);

        if(this.inBulkModification == 0)
            return;
        
        // XXX - should be accumulating modifications!!!
        final boolean modified = this.endBulkModificationImpl(successful);
        this.inBulkModification--;

        if((this.inBulkModification == 0) && modified)
            this.dispatchDataStoreContentChangedNoSync();
    }
    
    protected abstract boolean endBulkModificationImpl(boolean successful);
        
    @Override
    public final FeatureSet insertFeatureSet(final String provider, 
          final String type, final String name, 
          final double minResolution, final double maxResolution, 
          final boolean returnRef) {
        this.checkModificationFlags(MODIFY_FEATURESET_INSERT, true);
        
        synchronized(this) {
            FeatureSet[] ref = returnRef ? new FeatureSet[1] : null;
            if(this.insertFeatureSetImpl(provider, type, name, minResolution, maxResolution, ref))
                this.dispatchDataStoreContentChangedNoSync();
            return returnRef ? ref[0] : null;
        }
        
    }
    
    protected abstract boolean insertFeatureSetImpl(String provider, String type, String name, double minResolution, double maxResolution, FeatureSet[] returnRef);
    
    @Override
    public final void updateFeatureSet(long fsid, String name) {
        this.checkModificationFlags(MODIFY_FEATURESET_UPDATE|MODIFY_FEATURESET_NAME, true);
        
        synchronized(this) {
            if(this.updateFeatureSetImpl(fsid, name))
                this.dispatchDataStoreContentChangedNoSync();
        }
    }
    
    protected abstract boolean updateFeatureSetImpl(long fsid, String name);
    
    @Override
    public final void updateFeatureSet(long fsid, double minResolution, double maxResolution) {
        this.checkModificationFlags(MODIFY_FEATURESET_UPDATE|MODIFY_FEATURESET_DISPLAY_THRESHOLDS, true);
        
        synchronized(this) {
            if(this.updateFeatureSetImpl(fsid, minResolution, maxResolution))
                this.dispatchDataStoreContentChangedNoSync();
        }
    }
    
    protected abstract boolean updateFeatureSetImpl(long fsid, double minResolution, double maxResolution);
    
    @Override
    public final void updateFeatureSet(long fsid, String name, double minResolution, double maxResolution) {
        this.checkModificationFlags(MODIFY_FEATURESET_UPDATE|MODIFY_FEATURESET_NAME|MODIFY_FEATURESET_DISPLAY_THRESHOLDS, true);
        
        synchronized(this) {
            if(this.updateFeatureSetImpl(fsid, name, minResolution, maxResolution))
                this.dispatchDataStoreContentChangedNoSync();
        }
    }
    
    protected abstract boolean updateFeatureSetImpl(long fsid, String name, double minResolution, double maxResolution);
    
    @Override
    public final void deleteFeatureSet(long fsid) {
        this.checkModificationFlags(MODIFY_FEATURESET_DELETE, true);
        
        synchronized(this) {
            if(this.deleteFeatureSetImpl(fsid))
                this.dispatchDataStoreContentChangedNoSync();
        }
    }
    
    protected abstract boolean deleteFeatureSetImpl(long fsid);
    
    @Override
    public final void deleteAllFeatureSets() {
        this.checkModificationFlags(MODIFY_FEATURESET_DELETE, true);
        
        synchronized(this) {
            if(this.deleteAllFeatureSetsImpl())
                this.dispatchDataStoreContentChangedNoSync();
        }
    }
    
    protected abstract boolean deleteAllFeatureSetsImpl();
    
    @Override
    public final Feature insertFeature(final long fsid, final String name, final Geometry geom, final Style style, final AttributeSet attributes, final boolean returnRef) {
        this.checkModificationFlags(MODIFY_FEATURESET_FEATURE_INSERT, true);
        
        synchronized(this) {
            Feature[] ref = returnRef ? new Feature[1] : null;
            if(this.insertFeatureImpl(fsid, name, geom, style, attributes, ref))
                this.dispatchDataStoreContentChangedNoSync();
            return returnRef ? ref[0] : null;
        }
    }
    
    protected abstract boolean insertFeatureImpl(final long fsid, final String name, final Geometry geom, Style style, final AttributeSet attributes, final Feature[] returnRef);

    @Override
    public final Feature insertFeature(final long fsid, final FeatureDefinition def, final boolean returnRef) {
        this.checkModificationFlags(MODIFY_FEATURESET_FEATURE_INSERT, true);
        
        synchronized(this) {
            Feature[] ref = returnRef ? new Feature[1] : null;
            if(this.insertFeatureImpl(fsid, def, ref))
                this.dispatchDataStoreContentChangedNoSync();
            return returnRef ? ref[0] : null;
        }
    }
    
    protected abstract boolean insertFeatureImpl(long fsid, FeatureDefinition def, Feature[] returnRef);

    @Override
    public final void updateFeature(long fid, String name) {
        this.checkModificationFlags(MODIFY_FEATURESET_FEATURE_UPDATE|MODIFY_FEATURE_NAME, true);
        
        synchronized(this) {
            if(this.updateFeatureImpl(fid, name))
                this.dispatchDataStoreContentChangedNoSync();
        }
    }
    
    protected abstract boolean updateFeatureImpl(long fid, String name);
    
    @Override
    public void updateFeature(long fid, Geometry geom) {
        this.checkModificationFlags(MODIFY_FEATURESET_FEATURE_UPDATE|MODIFY_FEATURE_GEOMETRY, true);
        
        synchronized(this) {
            if(this.updateFeatureImpl(fid, geom))
                this.dispatchDataStoreContentChangedNoSync();
        }
    }
    
    protected abstract boolean updateFeatureImpl(long fid, Geometry geom);
    
    @Override
    public void updateFeature(long fid, Style style) {
        this.checkModificationFlags(MODIFY_FEATURESET_FEATURE_UPDATE|MODIFY_FEATURE_STYLE, true);
        
        synchronized(this) {
            if(this.updateFeatureImpl(fid, style))
                this.dispatchDataStoreContentChangedNoSync();
        }
    }
    
    protected abstract boolean updateFeatureImpl(long fid, Style style);
    
    @Override
    public final void updateFeature(long fid, AttributeSet attributes) {
        this.checkModificationFlags(MODIFY_FEATURESET_FEATURE_UPDATE|MODIFY_FEATURE_ATTRIBUTES, true);
        
        synchronized(this) {
            if(this.updateFeatureImpl(fid, attributes))
                this.dispatchDataStoreContentChangedNoSync();
        }
    }
    
    protected abstract boolean updateFeatureImpl(long fid, AttributeSet attributes);
    
    @Override
    public final void updateFeature(long fid, String name, Geometry geom, Style style, AttributeSet attributes) {
        this.checkModificationFlags(MODIFY_FEATURESET_FEATURE_UPDATE|MODIFY_FEATURE_NAME|MODIFY_FEATURE_GEOMETRY|MODIFY_FEATURE_STYLE|MODIFY_FEATURE_ATTRIBUTES, true);
        
        synchronized(this) {
            if(this.updateFeatureImpl(fid, name, geom, style, attributes))
                this.dispatchDataStoreContentChangedNoSync();
        }
    }
    
    protected abstract boolean updateFeatureImpl(long fid, String name, Geometry geom, Style style, AttributeSet attributes);
    
    @Override
    public final void deleteFeature(long fid) {
        this.checkModificationFlags(MODIFY_FEATURESET_FEATURE_DELETE, true);
        
        synchronized(this) {
            if(this.deleteFeatureImpl(fid))
                this.dispatchDataStoreContentChangedNoSync();
        }
    }
    
    protected abstract boolean deleteFeatureImpl(long fsid);
    
    @Override
    public final void deleteAllFeatures(long fsid) {
        this.checkModificationFlags(MODIFY_FEATURESET_FEATURE_DELETE, true);
        
        synchronized(this) {
            if(this.deleteAllFeaturesImpl(fsid))
                this.dispatchDataStoreContentChangedNoSync();
        }
    }
    
    protected abstract boolean deleteAllFeaturesImpl(long fsid);
    
    protected final boolean checkModificationFlags(int capability, boolean throwUnsupported) {
        final boolean retval = ((this.modificationFlags&capability) == capability);
        if(!retval && throwUnsupported)
            throw new UnsupportedOperationException();
        return retval;
    }

    /**************************************************************************/
    
    public static void setFeatureSetsVisible(FeatureDataStore dataStore, FeatureSetQueryParameters params, boolean visible) {
        LinkedList<Long> fsids = new LinkedList<Long>();

        FeatureSetCursor result = null;
        try {
            result = dataStore.queryFeatureSets(params);
            while(result.moveToNext())
                fsids.add(Long.valueOf(result.get().getId()));
        } finally {
            if(result != null)
                result.close();
        }
        
        for(Long fsid : fsids)
            dataStore.setFeatureSetVisible(fsid.longValue(), visible);
    }
    
    public static void setFeaturesVisible(FeatureDataStore dataStore, FeatureQueryParameters params, boolean visible) {
        LinkedList<Long> fids = new LinkedList<Long>();

        FeatureCursor result = null;
        try {
            result = dataStore.queryFeatures(params);
            while(result.moveToNext())
                fids.add(Long.valueOf(result.getId()));
        } finally {
            if(result != null)
                result.close();
        }
        
        for(Long fid : fids)
            dataStore.setFeatureVisible(fid.longValue(), visible);
    }
    
    public static int queryFeatureSetsCount(FeatureDataStore dataStore, FeatureSetQueryParameters params) {
        FeatureSetCursor result = null;
        try {
            result = dataStore.queryFeatureSets(params);
            int retval = 0;
            while(result.moveToNext())
                retval++;
            return retval;
        } finally {
            if(result != null)
                result.close();
        }
    }
    
    public static int queryFeaturesCount(FeatureDataStore dataStore, FeatureQueryParameters params) {
        FeatureCursor result = null;
        try {
            // since we're only counting the results, try to reduce the amount
            // of data we're pulling out of the datastore
            if(params != null)
                params = new FeatureQueryParameters(params);
            else
                params = new FeatureQueryParameters();
            
            params.ignoredFields = FeatureQueryParameters.FIELD_ATTRIBUTES |
                                   FeatureQueryParameters.FIELD_GEOMETRY |
                                   FeatureQueryParameters.FIELD_NAME |
                                   FeatureQueryParameters.FIELD_STYLE;

            result = dataStore.queryFeatures(params);
            final int limit = (params.limit+params.offset);
            int retval = 0;
            while(result.moveToNext()) {
                retval++;
                if(retval == limit)
                    break;
            }
            return Math.max(retval-params.offset, 0);
        } finally {
            if(result != null)
                result.close();
        }
    }
    
    public static Feature insertFeature(FeatureDataStore dataStore, long fsid, FeatureDefinition def, boolean returnRef) {
        final Feature feature = def.get();
        return dataStore.insertFeature(fsid,
                                       feature.getName(),
                                       feature.getGeometry(),
                                       feature.getStyle(),
                                       feature.getAttributes(),
                                       returnRef);
    }
 
    /**
     * Test to see if a value matches a test string containing optional wildcard characters.
     * @return true if the value matches the test, if the value is null, the method will return
     * false.
     */
    public static boolean matches(final String test, final String value, final char wildcard) {
        if (value == null)
             return false;

        if(test.indexOf(wildcard) < 0) 
            return value.equals(test);

        return value.matches(WildCard.wildcardAsRegex(test, wildcard));
    }

    public static boolean matches(Collection<String> test, String value, char wildcard) {
        for(String arg : test)
            if(matches(arg, value, wildcard))
                return true;
        return false;
    }
}
